% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/methods.R
\name{GenerateNull}
\alias{GenerateNull}
\title{Generate a list of null sites.}
\usage{
GenerateNull(locus, id = NULL, method = c("ntAbund", "perm"),
  nt = "C", showPb = TRUE)
}
\arguments{
\item{locus}{A \code{txLoc} object.}

\item{id}{A character string; identifier for null sites; if
\code{NULL} then id = \code{"null"}; default is \code{NULL}.}

\item{method}{A character string; specifies the method used
to generate null distribution; if \code{method == "ntAbund"}
the position of all nucleotides specified by \code{nt} will
be used as null sites; if \code{method == "perm"} then null
sites will be generated by uniform-randomly shuffling candidate
positions from \code{locus} within the corresponding transcript
region; default is \code{"ntAbund"}.}

\item{nt}{A single character; if \code{method == "ntAbund"},
use \code{nt} to derive distribution of null sites.}

\item{showPb}{A logical scalar; if \code{TRUE} show a progress
bar; default is \code{TRUE}.}
}
\value{
A \code{txLoc} object. See 'Details'.
}
\description{
Generate a list of null sites. See 'Details'.
}
\details{
The function generates a null distribution of single-nucleotide
sites across different transcript sections, and returns a 
\code{txLoc} object.
Two different methods can be employed:
\enumerate{
\item \code{method = "ntAbund"}: Null sites are generated based
on the position of all non-modified nucleotides of type \code{nt} 
in those transcript sections that also contain a modified site of
the same type and as specified in \code{locus}. 
For example, if \code{locus} contains a list of m$^{6}$A sites, 
the list of null sites consists of all non-methylated adenosines 
in those transcripts that contain at least one m$^{6}$A site.
\item \code{method = "perm"}: Null sites are generated by 
permuting the position of sites from \code{locus} uniformly within
the corresponding transcript section. Note that this will generate
a list of null sites with the same abundance ratios across 
transcript sections as the list of sites from \code{locus}. It is
therefore not useful for assessing an enrichment of sites within
a particular transcript section.
}
It is import to emphasise that any downstream enrichment analysis
may depend critically on the choice of the null distribution. 
For example, a position permution-based null distribution may not 
be a valid null distribution, if the nucleotide position 
distribution is highly non-uniform across a transcript section. 
This is the case e.g. for the spatial distribution of cytosines 
within and across the 5'UTR, CDS and/or 3'UTR. In this case, a
better null distribution would be to consider all cytosines
in transcript sections that also contain a site in \code{locus}.
This can be achieved with \code{method = "ntAbund"}. A still 
better list of null sites would be based on all \emph{expressed} 
and non-modified cytosines based on the same sequencing data that
was used to identify modified cytosines.
}
\examples{
\dontrun{
bedFile <- system.file("extdata",
                       "miCLIP_m6A_Linder2015_hg38.bed",
                       package = "RNAModR");
sites <- ReadBED(bedFile);
posSites <- SmartMap(sites, id = "m6A", refGenome = "hg38");
negSites <- GenerateNull(posSites, method = "ntAbund", nt = "A");
}

}
\author{
Maurits Evers, \email{maurits.evers@anu.edu.au}
}
